ignite/engine/utils.py

```diff
@@ -55,15 +55,31 @@ def __len__(self) -> int:
         return len(self.batch_sampler)
 
 
-def _check_signature(engine, fn: Callable, fn_description: str, *args, **kwargs) -> None:
+def _check_signature(fn: Callable, fn_description: str, *args, **kwargs) -> None:
 
     signature = inspect.signature(fn)
     try:
-        signature.bind(engine, *args, **kwargs)
+        signature.bind(*args, **kwargs)
     except TypeError as exc:
         fn_params = list(signature.parameters)
         exception_msg = str(exc)
-        passed_params = [engine] + list(args) + list(kwargs)
+        passed_params = list(args) + list(kwargs)
+        raise ValueError(
+            "Error adding {} '{}': "
+            "takes parameters {} but will be called with {} "
+            "({}).".format(fn, fn_description, fn_params, passed_params, exception_msg)
+        )
+
+
+def _check_partial_signature(fn: Callable, fn_description: str, *args, **kwargs) -> None:
```

Comment:
Could you please explain the need of partial bind ? 
I thought we could check both in a single `_check_signature` like method ?

Reply:
First, I think `_check_signature` could be refactor to have engine in `*args` and should be generic.

I implemented the `_check_partial_signature` to check arguments of an handler without knowing if it should contains the `engine`.

I should have preferred to check the signature of the handler wrt `*args` and `**kwargs` to see if engine is needed or not. And thus, engine should have been added to `*args`. No check needed at `_fire_event`.

But `_fire_event` could add own argument... And `engine` is always the first argument of `*args`.

Reply:
> I thought we could check both in a single `_check_signature` like method ?

We could do that 👍🏻

Reply:
Notice that with optional `engine`, the following snippet is correct
```python
def handler(arg):
    pass

engine.add_event_handler(Events.STARTED, handler)    # arg == engine 
engine.add_event_handler(Events.STARTED, handler, 1) # arg == 1
```

Reply:
Yes, it is correct. We need to put a similar example into docstring.